common/persistence/cassandra/cassandraTaskPersistence.go

```diff
@@ -201,350 +95,258 @@ func (d *cassandraTaskPersistence) LeaseTaskList(
+			return nil, &p.ConditionFailedError{
+				Msg: fmt.Sprintf("Failed to insert tasks. name: %v, type: %v, rangeID: %v, columns: (%v)",
+					request.TaskListInfo.Name, request.TaskListInfo.TaskType, request.TaskListInfo.RangeID, previous),
+			}
 		}
+		return nil, convertCommonErrors(t.db, "CreateTasks", err)
 	}
 
 	return &p.CreateTasksResponse{}, nil
 }
 
+func toTaskListRow(info *p.TaskListInfo) *nosqlplugin.TaskListRow {
+	return &nosqlplugin.TaskListRow{
+		DomainID:        info.DomainID,
+		TaskListName:    info.Name,
+		TaskListType:    info.TaskType,
+		TaskListKind:    info.Kind,
+		RangeID:         info.RangeID,
+		AckLevel:        info.AckLevel,
+		LastUpdatedTime: info.LastUpdated,
+	}
+}
+
 // From TaskManager interface
-func (d *cassandraTaskPersistence) GetTasks(
+func (t *nosqlTaskManager) GetTasks(
 	ctx context.Context,
 	request *p.GetTasksRequest,
 ) (*p.InternalGetTasksResponse, error) {
 	if request.MaxReadLevel == nil {
-		return nil, &types.InternalServiceError{
-			Message: "getTasks: both readLevel and maxReadLevel MUST be specified for cassandra persistence",
-		}
+		request.MaxReadLevel = common.Int64Ptr(math.MaxInt64)
 	}
+
 	if request.ReadLevel > *request.MaxReadLevel {
 		return &p.InternalGetTasksResponse{}, nil
 	}
 
-	// Reading tasklist tasks need to be quorum level consistent, otherwise we could loose task
-	query := d.session.Query(templateGetTasksQuery,
-		request.DomainID,
-		request.TaskList,
-		request.TaskType,
-		rowTypeTask,
-		request.ReadLevel,
-		*request.MaxReadLevel,
-	).PageSize(request.BatchSize).WithContext(ctx)
-
-	iter := query.Iter()
-	if iter == nil {
-		return nil, &types.InternalServiceError{
-			Message: "GetTasks operation failed.  Not able to create query iterator.",
-		}
-	}
+	resp, err := t.db.SelectTasks(ctx, &nosqlplugin.TasksFilter{
+		TaskListFilter: nosqlplugin.TaskListFilter{
+			DomainID:     request.DomainID,
+			TaskListName: request.TaskList,
+			TaskListType: request.TaskType,
+		},
+		BatchSize: request.BatchSize,
 
-	response := &p.InternalGetTasksResponse{}
-	task := make(map[string]interface{})
-PopulateTasks:
-	for iter.MapScan(task) {
-		taskID, ok := task["task_id"]
-		if !ok { // no tasks, but static column record returned
-			continue
-		}
-		t := createTaskInfo(task["task"].(map[string]interface{}))
-		t.TaskID = taskID.(int64)
-		response.Tasks = append(response.Tasks, t)
-		if len(response.Tasks) == request.BatchSize {
-			break PopulateTasks
-		}
-		task = make(map[string]interface{}) // Reinitialize map as initialized fails on unmarshalling
+		MinTaskID: request.ReadLevel,
+		MaxTaskID: *request.MaxReadLevel,
+	})
+
+	if err != nil {
+		return nil, convertCommonErrors(t.db, "GetTasks", err)
 	}
 
-	if err := iter.Close(); err != nil {
-		return nil, convertCommonErrors(d.client, "GetTasks", err)
+	response := &p.InternalGetTasksResponse{}
+	for _, t := range resp {
+		response.Tasks = append(response.Tasks, toTaskInfo(t))
```

Comment:
>Some random thoughts: looks like we can just have one rangeDelete API in the plugin interface so that development don't need to implement both delete and rangeDelete?

Ha, that's a genius idea! Will do it. 

>  there's no nextPageToken in the response even when batchSize in the request is less than the number of tasks in the requested range
yeah, the way it works it that in tasklist scanner, the activity will run again and delete next page.